### ISSUE TITLE
Add calculation of @DEBYETEMPERATURE from VDOS at different temperatures

### ISSUE THREAD

**Issue by ramic-k** [mctools/ncrystal#58]
Hello,
It would be useful to be able to calculate the temperature dependent @DEBYETEMPERATURE from VDOS, instead of using one constant @DEBYETEMPERATURE for all temperatures. 

This can be done by calculating MSD from VDOS and then using msd (in proper units) one can obtain DebyeTemperature using **debyeTempFromIsotropicMSD**(double msd, double temperature, double atomic_mass) from internal class **NCDebyeMSD.hh**.

We can calculate the MSD from VDOS using (in this case using python syntax and methods):
k = 25.3e-3/293.6 # eV/K
hbar = 0.6582e-3 # eV\*ps
amu = 0.103642697e-3 # eV\*ps^2\*A^-2
kT = k\*Temperature
func = rho/(erg\*np.tanh(erg/(2\*kT)))
msd_from_vdos = hbar\*\*2/(2\*amu\*M)\*np.trapz(func, erg),

where the Temperature is the desired temperature; rho and erg are the VDOS values and the energy grid for the VDOS values; np.tanh is the numpy implementation of hyperbolic tangent function, M is the atomic mass in amu of the element whose VDOS is being used; and np.trapz is the numpy implementation of the numerical trapezoidal integration. (*It is important to note that below the first point in (erg,rho) parabolic behavior should be forced all the way to some small value near 0, so that the MSD can be calculated correctly.) 
The plots of the spread of msd extracted from all .NCMAT, files that have VDOS in them, versus msd_from_vdos calculated using above suggested method can be seen in:
[LogMSDvsLogMSDfromVDOS_temperature dependence.pdf](https://github.com/mctools/ncrystal/files/5970721/LogMSDvsLogMSDfromVDOS_temperature.dependence.pdf)
As it can be seen there is definitely a difference between the proposed method and using constant msd for all temperatures. 

To check the new proposed method we used the calculated msd with **debyeTempFromIsotropicMSD** from **NCDebyeMSD.hh** to obtain the new temperature dependent @DEBYETEMPERATURE and ran that to compare how the two methods calculate the scattering cross sections for all the .NCMAT files that had VDOS in them. We did this for the same temperatures: 20, 296 and 600 K. The results can be seen in (one of the easiest examples to look at is W):
[interpolatedVDOS_DebyeTempFromMSDfromVDOS_scattering_cross_section_plots.pdf](https://github.com/mctools/ncrystal/files/5970777/interpolatedVDOS_DebyeTempFromMSDfromVDOS_scattering_cross_section_plots.pdf)

The results for the msd spread combined with the results for the scattering cross sections suggest that the proposed method would give a more consistent method to calculate the different parts of the scattering cross sections for materials or .NCMAT files that use VDOS.

Regards,
Kemal and @marquezj.

**Comment by tkittel** [comment]
@ramic-k 

Hi Kemal,

Thanks a lot for summarising your work and our discussions here! As I told you, your plots clearly show that this is something we should adopt into NCrystal (with the NCVDOSEval class performing the integrals, since it already has the machinary for doing so with high accuracy - including the parabolic DOS behaviour at low frequencies - all the way to 0). It will certainly make the behaviour at very low or very high temperatures much more trustworthy.

Could you please post here also some references for the specific formula (the LEAPR manual?), so we can make it clear?

And just to clarify, when you write "constant msd at all temperatures", I think you meant "constant Debye temperature at all temperatures". Because our msd is derived from both the Debye temperature and the temperature, and is thus already temperature dependent. It just seems that integrating the actual vdos with the LEAPR(?) formula gives a better msd estimate at each T than estimating msd from a fixed debye temperature :-)

Given the incoming mega-release (to be named NCrystal v2.5.0 most likely), it will probably take around O(1month) for this feature to make it into an actual NCrystal release. At that time, you can generate your pdf again to verify that your implementations agree.

Cheers,
Thomas

**Comment by tkittel** [comment]
@ramic-k or @marquezj  can you please provide the requested info:

> Could you please post here also some references for the specific formula (the LEAPR manual?), so we can make it clear?

Then I might be able to add some preliminary support in the next release, but it requires me to verify the formula and also know exactly what reference to put in the documentations (I have learned the hard way that these references always have to be put in from the beginning).

**Comment by tkittel** [comment]
Great, so it was there all along :-)

Did you actually check that the equations are the same, or did you just look at the numbers coming out of it?

**Comment by tkittel** [comment]
And I am a bit surprised actually, I am integrating vdos(E)/(E*tanh(E/kT)) but I remember Kemal showed me a formula with vdos(E)*E/(tanh(E/kT)). Can one of you send me a link to wherever it was Kemal had the formula? I would like to double-check that all is OK, and it is not just some accidental result valid only in certain scenarios.

**Comment by marquezj** [comment]
There was probably a typo. The NJOY manual is [here](https://github.com/njoy/NJOY2016-manual/raw/master/njoy16.pdf). Basically the integral is eq. 526 where the integral from -inf to inf of the exponential makes a cosh(beta/2) factor. The P(beta) function there is defined in eq. 512 as rho(beta)/(beta*sinh(beta/2)), and from those two you get the same 1/(E*tanh(E/2kT) factor as in NCrystal. In practice, it is computed in [fsum()](https://github.com/njoy/NJOY2016/blob/e722511e88e4568502b51a15cebd38d03fd060a8/src/leapr.f90#L691), which is called [here](https://github.com/njoy/NJOY2016/blob/e722511e88e4568502b51a15cebd38d03fd060a8/src/leapr.f90#L657) and you can see that it implements that equation.

**Comment by tkittel** [comment]
Excellent, thanks for double-checking!

**Comment by tkittel** [comment]
So I just made a commit in our local repo (will only appear on github in ~1 week), where I added a preliminary way to use this. In order to postpone the NCMAT format evolution, you have to explicitly enable the feature by adding to your .ncmat file (which must be at least NCMAT v3):

```
@CUSTOM_TMPHACKPREFERMSDFROMVDOS
```
I.e. add an (empty) section with the name as shown at the end of the file, and NCrystal will calculate MSD from the VDOS and use that instead of the Debye temperature. Although hacky, this is better than using environment variables since using (and changing) environment variables tricks NCrystal's caching mechanism. It also has the advantage of working equally well from Python and C++.

You can of course add the lines (and change to NCMAT v3) in-memory only (using NC.getFileContent and NC.registerInMemoryFileData), so it is not as unwieldy as it sounds. In fact, I did just that in order to produce the comparison plots for all VDOS-enabled materials in our library with a small python script:

[vdosmsd_crystalswithvdos.pdf](https://github.com/mctools/ncrystal/files/6042874/vdosmsd_crystalswithvdos.pdf)

As a crazy idea, I was wondering if feeding the idealised parabolic vdos based on the Debye temperature into the same machinary, would provide a better temperature dependency even in the scenario without an actual VDOS. Sadly, it doesn't really make any difference at all. I guess there is more mathematical consistency behind the models than I can spot:

[vdosmsd_crystalswithoutvdos.pdf](https://github.com/mctools/ncrystal/files/6042873/vdosmsd_crystalswithoutvdos.pdf)

So in summary: in O(1week) you will have a short-term way to use this feature and in O(1month) it will be available properly in an actual NCrystal release. At that point we should probably remove the Debye temperatures from all official ncmat files for which VDOS are available (and bump them to the new NCMAT v4).

**Comment by tkittel** [comment]
I forgot to mention that the plots are shown both vs. wavelength (first half of pdf) and vs. energy (second half), and the breakdown in components are shown as well. We can see that most of the effect is mainly due to the inclination of the bragg cross section at higher energies.

**Comment by ramic-k** [comment]
Thank you Thomas, looking forward to using it, because that will cut down the length of our NJOY C++ wrapper significantly!

**Comment by tkittel** [comment]
As you know, v2.6.0 should be released within the next few days. For completeness, I attach here a plot created with this new release, showing how Debye temperatures and mean-squared-displacements vary as a function of temperature, for a) everything derived directly from the VDOS, and b) everything derived from a fixed Debye temperature:

[msd_debyetemp_vdos_versus_hardcoded.pdf](https://github.com/mctools/ncrystal/files/6254972/msd_debyetemp_vdos_versus_hardcoded.pdf)

The plots are pretty interesting, they actually show remarkably similar predictions for the two methods, except of course for the low-temperature divergences for a few materials.

**Comment by tkittel** [comment]
Release 2.6.0 is out, closing here.

https://github.com/mctools/ncrystal/releases/tag/v2.6.0
